render/drm_dumb_allocator.c

```diff
@@ -0,0 +1,183 @@
+#include "render/drm_dumb_allocator.h"
+
+#include <wlr/util/log.h>
+
+#include <xf86drm.h>
+#include <xf86drmMode.h>
+
+#include <unistd.h>
+#include <assert.h>
+#include <stdlib.h>
+#include <string.h>
+#include <sys/mman.h>
+
+static const struct wlr_buffer_impl buffer_impl;
+
+static struct wlr_drm_dumb_buffer *drm_dumb_buffer_from_buffer(
+		struct wlr_buffer *wlr_buf) {
+	assert(wlr_buf->impl == &buffer_impl);
+	return (struct wlr_drm_dumb_buffer *)wlr_buf;
+}
+
+static void finish_buffer(struct wlr_drm_dumb_buffer *buf) {
+
+	if (buf->data) {
+		munmap(buf->data, buf->size);
+	}
+
+	if (buf->drm_fd >= -1) {
+		struct drm_mode_destroy_dumb destroy = { .handle = buf->handles[0] };
+		drmIoctl(buf->drm_fd, DRM_IOCTL_MODE_DESTROY_DUMB, &destroy);
+	}
+
+	wl_list_remove(&buf->link);
+}
+
+static struct wlr_drm_dumb_buffer *create_buffer(
+		struct wlr_drm_dumb_allocator *alloc, int width, int height,
+		const struct wlr_drm_format *format) {
+	struct wlr_drm_dumb_buffer *buffer = calloc(1, sizeof(*buffer));
+	if (buffer == NULL) {
+		return NULL;
+	}
+	wlr_buffer_init(&buffer->base, &buffer_impl, width, height);
+	wl_list_insert(&alloc->buffers, &buffer->link);
+
+	struct drm_mode_create_dumb create = {0};
+	create.width = (uint32_t)width;
+	create.height = (uint32_t)height;
+	create.bpp = 32;
```

Comment:
We should probably only support ARGB8888 and XRGB8888 here.

Reply:
So this should be a constant somewhere and the format should be checked right? Would it be possible to have a different bpp value with other formats?

Reply:
Yeah, there's a direct mapping between DRM formats and bpp for dumb buffers. 32 means ARGB8888 and 24 means XRGB8888. A simple switch is probably enough.

Reply:
Should I make it futureproof? Or we just wont support any other format

Reply:
As discussed on IRC, ARGB8888 and XRGB8888 both have a BPP of 32, if the format requested is not one of those, the creation of the buffer should fail gracefully.

Reply:
Done.